compiler/src/iree/compiler/Codegen/Common/SpecializeExports.cpp

```diff
@@ -462,6 +463,18 @@ struct SpecializeExportsPass final
       specializeExportedFunctionByRangeAttribute(exportOp, exportedFunc, helper,
                                                  ordinalSet);
     }
+
+    // TODO(#21623): We need DCE after this pass as it can leave around
+    // `tensor.dim` operations that can mess up the next passes (e.g.
+    // MaterializeDeviceEncoding). Ideally, we would avoid creating
+    // those ops altogether if not needed.
+    MLIRContext *ctx = &getContext();
+    RewritePatternSet patterns(ctx);
+    ctx->getLoadedDialect<tensor::TensorDialect>()->getCanonicalizationPatterns(
```

Comment:
Is there a narrower set of patterns we can use? I mean for specific `tensor` ops.

Reply:
Yes, I limited it to just the ones for `tensor.dim` now.